recipes/lora_finetune_single_device.py

```diff
@@ -557,31 +576,10 @@ def train(self) -> None:
                     ):
                         break
 
-                    # Both are shape [b, s]
-                    tokens, labels = batch["tokens"], batch["labels"]
-                    # Get the attention mask and position ids from the dataset if they
-                    # exist. Currently, only sample packing in PackedDataset returns these
-                    mask = batch.get("mask", None)  # shape [b, s, s]
-                    input_pos = batch.get("input_pos", None)  # shape [b, s]
-
-                    tokens = tokens.to(self._device)
-                    num_tokens += tokens.numel()
-                    labels = labels.to(self._device)
-                    mask = mask.to(self._device) if mask is not None else None
-                    input_pos = (
-                        input_pos.to(self._device) if input_pos is not None else None
-                    )
-
-                    logits = self._model(tokens, mask=mask, input_pos=input_pos)
-                    # Shift so that tokens < n predict n
-                    logits = logits[..., :-1, :].contiguous()
-                    labels = labels[..., 1:].contiguous()
-                    logits = logits.transpose(1, 2)
-                    # Compute loss
-                    loss = self._loss_fn(logits, labels)
-                    # free logits otherwise it peaks backward memory
-                    del logits
+                    batch = {k: v.to(self._device) for k, v in batch.items()}
```

Comment:
My concern is that the batch may contain keys that may not be necessary in the gpu. Can we keep the .to(device) like it was before? Or did you have some other motivation for it?

Reply:
I tried putting the old `.to(device)` code within `self._loss_step()`, but `torch.compile()` shows some warnings. Should be harmless though. I will double-check it again.

Another option is to keep the old `.to(device)` outside `self._loss_step()`. Then, it is a bit more clunky to pass `tokens`, `labels`, `mask`, `input_pos` to `self._loss_step()` instead of just `batch` (or we can just modify the `batch` object in-place). Lmk which one you prefer, then I will just change to that.

Reply:
i think it should be fine to keep as it is then. I dont think that our batches output extra terms, so that scenario I raised would be an exception.